src/java.base/windows/classes/sun/nio/fs/WindowsFileSystemProvider.java

```diff
@@ -418,65 +419,137 @@ public void checkAccess(Path obj, AccessMode... modes) throws IOException {
         }
     }
 
-    @Override
-    public boolean isSameFile(Path obj1, Path obj2) throws IOException {
-        WindowsPath file1 = WindowsPath.toWindowsPath(obj1);
-        if (file1.equals(obj2))
-            return true;
-        if (obj2 == null)
-            throw new NullPointerException();
-        if (!(obj2 instanceof WindowsPath))
+    // contains file attributes and its handle
+    // the attributes' key is valid as long as the handle remains open
+    private record LinkAttributes(WindowsFileAttributes attrs, long handle) {
+        public boolean equals(Object obj) {
+            if (obj == this)
+                return true;
+            if (obj instanceof LinkAttributes other)
+                return WindowsFileAttributes.isSameFile(attrs, other.attrs());
```

Comment:
It's not obvious that this consistent with hashCode. It might be better to have this check the that the 3 values are equal.